backend/s3: add `use_legacy_workflow` argument to `modernization` branch

jar-b opened the pull request:
<!--

Describe in detail the changes you are proposing, and the rationale.

See the contributing guide:

https://github.com/hashicorp/terraform/blob/main/.github/CONTRIBUTING.md

-->
Adds an optional `use_legacy_workflow` argument to preserve the authentication flow `aws-sdk-go-base` V1 implemented. This does not align with the flow of the AWS CLI or SDK's and will be phased out over several minor releases. In `v1.6.0` this argument will default to `true` (ie. no user facing changes).

```console
% TF_ACC=1 go test -count=1 ./internal/backend/remote-state/s3/...
ok      github.com/hashicorp/terraform/internal/backend/remote-state/s3 103.779s
```


<!--

Link all GitHub issues fixed by this PR, and add references to prior
related PRs.

-->

Relates #33631
Relates #30443
Relates https://github.com/hashicorp/aws-sdk-go-base/pull/595

## Target Release

<!--

In normal circumstances we only target changes at the upcoming minor
release, or as a patch to the current minor version. If you need to
port a security fix to an older release, highlight this here by listing
all targeted releases.

If targeting the next patch release, also add the relevant x.y-backport
label to enable the backport bot.

-->

1.6.0

gdavison commented on internal/backend/remote-state/s3/backend.go:
```diff
@@ -230,6 +230,13 @@ func (b *Backend) ConfigSchema() *configschema.Block {
+				Type:        cty.Bool,
+				Optional:    true,
+				Description: "Use the legacy authentication workflow, preferring environment variables over backend configuration.",
+				Deprecated:  true,
```
Since we won't be removing it until v1.8, we can probably hold off on deprecating it until v1.7

gdavison reviewed (approved):
One comment, otherwise looks good 🚀

jar-b commented:
Removed the deprecation and added one additional test case to cover an explicitly disabled legacy workflow.

```console
% TF_ACC=1 go test -count=1 ./internal/backend/remote-state/s3/...
ok      github.com/hashicorp/terraform/internal/backend/remote-state/s3 98.150s
```
